core/embed/rust/src/trezorhal/storage.rs

```diff
@@ -0,0 +1,322 @@
+}
+
+pub fn ensure_not_wipe_pin(pin: &str) {
+    unsafe {
+        ffi::storage_ensure_not_wipe_code(pin.as_ptr(), pin.len());
+    }
+}
+
+/// Check if value for `appkey` exists.
+pub fn has(appkey: u16) -> bool {
+    ffi::sectrue == unsafe { ffi::storage_has(appkey) }
+}
+
+/// Get length of value stored for `appkey`.
+/// Returns an error if the value does not exist.
+pub fn get_length(appkey: u16) -> StorageResult<usize> {
+    let mut len = 0;
+    if ffi::sectrue == unsafe { ffi::storage_get(appkey, ptr::null_mut() as _, 0, &mut len) } {
+        Ok(len as usize)
+    } else {
+        Err(StorageError::ReadFailed)
+    }
+}
+
+/// Get value for `appkey` from storage.
+/// The provided `data` buffer must be long enough to hold the value. If the
+/// buffer is shorter than the stored value, an error is returned. Use
+/// `get_length` to find the size of the value beforehand. On success, the data
+/// is copied to `data`, and length of copied data is returned. Data previously
+/// in the buffer beyond the copied data is not modified.
+/// If `appkey` is not present in storage, or the value is private and storage
+/// is locked, returns an error.
+pub fn get(appkey: u16, data: &mut [u8]) -> StorageResult<usize> {
+    let max_len = data.len().min(u16::MAX as usize) as u16;
+    let mut len = 0u16;
+    if ffi::sectrue
+        == unsafe { ffi::storage_get(appkey, data.as_mut_ptr() as _, max_len, &mut len) }
+    {
+        Ok(len as usize)
+    } else {
+        Err(StorageError::ReadFailed)
+    }
+}
+
+/// Set value for `appkey` in storage.
+/// The maximum length of the value is `u16::MAX`. Passing longer buffers will
+/// result in an error.
+/// If storage is locked and the value is not public-writable, returns an error.
+pub fn set(appkey: u16, data: &[u8]) -> StorageResult<()> {
+    if data.len() > u16::MAX as usize {
+        Err(StorageError::InvalidData)
+    } else if ffi::sectrue
+        == unsafe { ffi::storage_set(appkey, data.as_ptr() as _, data.len() as u16) }
+    {
+        Ok(())
+    } else {
+        Err(StorageError::WriteFailed)
+    }
+}
+
+/// Set value for monotonic counter `appkey` in storage.
+pub fn set_counter(appkey: u16, counter: u32) -> StorageResult<()> {
+    if ffi::sectrue == unsafe { ffi::storage_set_counter(appkey, counter) } {
+        Ok(())
+    } else {
+        Err(StorageError::WriteFailed)
+    }
+}
+
+/// Atomically get and increment the value for monotonic counter `appkey` in
+/// storage.
+pub fn get_next_counter(appkey: u16) -> StorageResult<u32> {
+    let mut counter = 0u32;
+    if ffi::sectrue == unsafe { ffi::storage_next_counter(appkey, &mut counter) } {
+        Ok(counter)
+    } else {
+        Err(StorageError::CounterFailed)
+    }
+}
+
+/// Delete value for `appkey` from storage.
+/// Returns true if the value was successfully deleted.
+/// If the storage is locked, or the value does not exist in storage, returns
+/// false.
+pub fn delete(appkey: u16) -> bool {
```

Comment:
Why not `StorageResult<()>` like elsewhere?

Reply:
i think because the `false` result is not necessarily an "error"?

Reply:
still the logic is similar to `get`, where locked storage or non present key returns an error

Reply:
Similar but not the same:
For `get`, the error `ReadFailed` indicates that we didn't get anything. This is an error in any case.
For `delete`, an error would mean that either the value isn't there (in which case it could be considered "success"), or we failed to delete it -- because storage is locked or because of i/o error.

But for API coherence, it seems more reasonable to declare and document that trying to delete nonexistent key _is an error_ (and then the higher level `storage_field` api will need to explicitly check if `storage.has(key)` before trying to delete it)

Reply:
https://github.com/trezor/trezor-firmware/commit/f5839559c763ee55cd6d1e23bf0385da2dbdb02d